File: hyper_expo.py
#coding: utf8
"""
	Script Python simulant le comportement de l'hyper exponentielle.
"""

from math import exp
import random

PROBABILITE_BURST = 0.01
NOMBRE_MESSAGE_BURST = 140
LAMBDA = 1
TAILLE_TABLEAU = 10
global TABLEAU_POISSON


def effectuer_tirage(probabilite):
	tirage = random.uniform(0, 1)
	print "Tirage : " ,tirage
	return tirage <= probabilite


def hyper_expo():
	if effectuer_tirage(PROBABILITE_BURST) == True:		#Le tirage est tombé sur la faible proba
		print "Burst !\nOn envoie : ", NOMBRE_MESSAGE_BURST, " messages."
	else:
		u = random.uniform(0, 1)
		print "Resultat loi de poisson via l'algo naif : ", loi_de_poisson_naif(u)
		print "Resultat loi de poisson via l'algo optimisé : ", loi_de_poisson_opti(u)


def loi_de_poisson_naif(u):
	p = exp (- LAMBDA)
	x = 0
	f = p
	while (u > f):
		x += 1
		p = p*LAMBDA/x
		f += p
	return x


def loi_de_poisson_opti(u):
	#Initialisation des variables
	initialiser_tableau()

	maxi = TAILLE_TABLEAU
	maxF = TABLEAU_POISSON[ TAILLE_TABLEAU -1]
	maxP = maxF

	f = TABLEAU_POISSON

	if u <= maxF:		#On peut trouver la valeur à l'aide du tableau
		x = 0
		while u > f[x]:
			x += 1
	else:
		x = maxi
		f = maxF
		p = maxP

		while u > f:
			x += 1
			p = p*LAMBDA/x
			f = f + p

	return x

def initialiser_tableau():
	global TABLEAU_POISSON
	TABLEAU_POISSON = [0] * TAILLE_TABLEAU

	TABLEAU_POISSON[0] = exp(-LAMBDA)
	p = TABLEAU_POISSON[0]

	for i in range(1, TAILLE_TABLEAU):
		p = p*LAMBDA/i
		TABLEAU_POISSON[i] = TABLEAU_POISSON[i-1]+p


def afficher_tableau():
	for i in range(TAILLE_TABLEAU):
		print "Indice : ", i
		print "valeur : ", TABLEAU_POISSON[i], "\n"


#####################     M A I N     #######################

hyper_expo()

File: simulation_calcul/Sources/hyper_expo.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include "../Headers/hyper_expo.h"

int effectuer_tirage(float probabilite)
{
	float tirage = generer_aleatoire(0, 1.00f);
	return tirage <= probabilite;
}

float generer_aleatoire(float borne_min, float borne_max)
{
	float tirage = ((float)rand() / RAND_MAX * (borne_max - borne_min) ) + borne_min;
	return tirage;
}

int loi_de_poisson_naif(float u, float lambda)
{
	double p = exp (- lambda);
	int x = 0;
	double f = p;
	do
	{
		x += 1;
		p = p*lambda/x;
		f += p;
	} while (u > f);
	return x;
}

int loi_de_poisson_opti(float u, float lambda, TableauDynamiqueDouble *tableau_poisson)
{
	int maxi = tableau_poisson->taille_utilisee;
	double *tableau = tableau_poisson->tableau;
	double maxF = tableau[ maxi -1];
	double maxP = maxF;

	int x;
	if (u <= maxF)		//On peut trouver la valeur à l'aide du tableau
	{
		x = 0;
		do
		{
			x++;
		} while (u > tableau[x-1]);
	}
	else
	{
		x = maxi;
		double f = maxF;
		double p = maxP;

		while (u > f)
		{
			x++;
			p = p*lambda/x;
			f = f + p;
		}
	}
	return x;
}

TableauDynamiqueDouble *initialiser_tableau_poisson(float lambda)
{
	TableauDynamiqueDouble *tableau_poisson = initialiser_tableau_dynamique_double();
	ajouter_valeur_tableau_dynamique_double(tableau_poisson, exp(-lambda) );

	double p = tableau_poisson->tableau[0];
	double seuil = 0.9999999;
	double somme = 0;

	int compteur = 1;
	while (somme < seuil)
	{
		p = p*lambda/compteur;
		double nouvelle_valeur = tableau_poisson->tableau[ compteur-1 ]+p;
		ajouter_valeur_tableau_dynamique_double(tableau_poisson, nouvelle_valeur);
		somme = tableau_poisson->tableau[ compteur ];
		compteur++;
	}

	return tableau_poisson;
}

int hyper_expo(TableauDynamiqueDouble *tableau_poisson)
{
	float u = generer_aleatoire(0, 1.0f);
	if (effectuer_tirage(PROBABILITE_BURST) == 1)	//Lambda grand
	{
		int message =  loi_de_poisson_naif(u, LAMBDA_GRAND);
		printf("Burst : %d\n", message);
		return message;
	}
	else											//lambda petit
	{
		int message =  (loi_de_poisson_opti(u, LAMBDA_PETIT, tableau_poisson));
		printf("Normal : %d\n", message);
		return message;
	}
}
